File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/simple/HiddenSingles.kt
package sudokusolver.kotlin.logic.simple

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.SetValue
import sudokusolver.kotlin.SudokuNumber
import sudokusolver.kotlin.UnsolvedCell

/*
 * http://www.sudokuwiki.org/Getting_Started
 *
 * If a candidate exists in only one cell in a unit, then the candidate must be placed in that cell.
 */
fun hiddenSingles(board: Board<Cell>): List<SetValue> =
    board.units.flatMap { unit ->
        val unsolved = unit.filterIsInstance<UnsolvedCell>()
        SudokuNumber.values().mapNotNull { candidate ->
            unsolved.singleOrNull { candidate in it.candidates }?.let { SetValue(it, candidate) }
        }
    }.distinct()

File: sudokusolver_scala/src/main/scala/sudokusolver/scala/logic/simple/HiddenSingles.scala
package sudokusolver.scala.logic.simple

import sudokusolver.scala.*

/*
 * http://www.sudokuwiki.org/Getting_Started
 *
 * If a candidate exists in only one cell in a unit, then the candidate must be placed in that cell.
 */
def hiddenSingles(board: Board[Cell]): Seq[SetValue] =
  board.units.flatMap { unit =>
    val unsolved = unit.collect { case cell: UnsolvedCell => cell }
    SudokuNumber.values.flatMap { candidate =>
      unsolved.filter(_.candidates.contains(candidate)) match
        case Seq(cell) => Some(SetValue(cell, candidate))
        case _ => None
    }
  }.distinct